<!DOCTYPE html>
<html>
<head>
<meta http-equiv="Content-Type" content="text/html; charset=utf-8"/>

<title>Why ReFernment</title>

<script type="text/javascript">
window.onload = function() {
  var imgs = document.getElementsByTagName('img'), i, img;
  for (i = 0; i < imgs.length; i++) {
    img = imgs[i];
    // center an image if it is the only element of its parent
    if (img.parentElement.childElementCount === 1)
      img.parentElement.style.textAlign = 'center';
  }
};
</script>





<style type="text/css">
body, td {
   font-family: sans-serif;
   background-color: white;
   font-size: 13px;
}

body {
  max-width: 800px;
  margin: auto;
  padding: 1em;
  line-height: 20px;
}

tt, code, pre {
   font-family: 'DejaVu Sans Mono', 'Droid Sans Mono', 'Lucida Console', Consolas, Monaco, monospace;
}

h1 {
   font-size:2.2em;
}

h2 {
   font-size:1.8em;
}

h3 {
   font-size:1.4em;
}

h4 {
   font-size:1.0em;
}

h5 {
   font-size:0.9em;
}

h6 {
   font-size:0.8em;
}

a:visited {
   color: rgb(50%, 0%, 50%);
}

pre, img {
  max-width: 100%;
}
pre {
  overflow-x: auto;
}
pre code {
   display: block; padding: 0.5em;
}

code {
  font-size: 92%;
  border: 1px solid #ccc;
}

code[class] {
  background-color: #F8F8F8;
}

table, td, th {
  border: none;
}

blockquote {
   color:#666666;
   margin:0;
   padding-left: 1em;
   border-left: 0.5em #EEE solid;
}

hr {
   height: 0px;
   border-bottom: none;
   border-top-width: thin;
   border-top-style: dotted;
   border-top-color: #999999;
}

@media print {
   * {
      background: transparent !important;
      color: black !important;
      filter:none !important;
      -ms-filter: none !important;
   }

   body {
      font-size:12pt;
      max-width:100%;
   }

   a, a:visited {
      text-decoration: underline;
   }

   hr {
      visibility: hidden;
      page-break-before: always;
   }

   pre, blockquote {
      padding-right: 1em;
      page-break-inside: avoid;
   }

   tr, img {
      page-break-inside: avoid;
   }

   img {
      max-width: 100% !important;
   }

   @page :left {
      margin: 15mm 20mm 15mm 10mm;
   }

   @page :right {
      margin: 15mm 10mm 15mm 20mm;
   }

   p, h2, h3 {
      orphans: 3; widows: 3;
   }

   h2, h3 {
      page-break-after: avoid;
   }
}
</style>



</head>

<body>
<h2>Why ReFernment</h2>

<p>ReFernment was created in response to the drudgery of manually annotating a plastome sequence that has levels of RNA editing. These editing sites result in a 
genomic sequence that contains features that look like errors (e.g., internal stop codons) so the annotations will be rejected by NCBI unless they are annotated 
correctly. ReFernment corrects the annotation to produce files ready for GenBank submission. ReFernment takes as input a GenBank flat file (.gb) and a GFF3 file 
(no sequence) and generates new annotations for nonsense mutations that can reasonably be explained by RNA editing, and provides conceptual translations for 
coding sequences with RNA editing. It should be made clear that ReFernment is not intended to predict every RNA editing site, as can other available software 
packages. These packages rely on RNA-Seq data to compare the genomic DNA, and thus determine if a nucleotide has been edited, but these data are not always 
available to researchers. We made ReFernment as a simple tool to save time for those who donâ€™t have RNA-Seq data available to them. We hope that ReFernment also 
saves the time of GenBank staff, as well has helping improve the quality of annotations available in GenBank.</p>

<h2>How it works</h2>

<p>ReFernment operates by refining existing annotations. Thus, the software uses an annotation generated by programs such as DOGMA, CpGAVAS, Verdant or AGORA 
(Wyman et al., 2004; Liu et al., 2012; McKain et al., 2017; Jung et al., 2018), and adjusts these annotations to account for RNA editing. ReFernment requires 
both a GFF3 (no sequence) file and a GenBank flat file (including nucleotide sequence) and its basic operation is extremely simple. First, ReFernment checks the
start and stop codons of each gene. In both cases ReFernment initially checks whether the codon is a valid start or stop, if the codon is not valid it checks 
whether an RNA editing event would result in the restoration of the codon to a valid start or stop (e.g. ACG -&gt; AUG). If the codon is not valid, even after checking 
for possible RNA editing, ReFernment checks whether nearby codons (within 5 nucleotides) represent valid codons; if so, refernment changes the gene boundaries to 
start or stop at those valid sites. Next, ReFernment checks whether a gene has any internal stops, and if so, checks whether RNA editing would restore these nonsense 
translations, adjusting the translation to account for this. ReFernment then edits the imputed GenBank flat file, adding conceptual translations and annotations 
indicating the sites where RNA editing occurred with â€˜misc_featureâ€™ flags, adding necessary RNA editing flags to the relevant genes, and providing a conceptual 
translation for each gene. Finally, ReFernment produces a five column feature table, formatted correctly for submission to GenBank, and a protein fasta file with the 
conceptual translations for coding sequences where RNA editing has occurred.</p>

<p>ReFernment operates under the assumption that only U-to-C or C-to-U RNA editing is occurring in the plastome (Takenaka et al., 2013) Additionally, ReFernment assumes 
that all nonsense mutations are the result of RNA editing. Since most of the genes that reside within the plastome are vital to photosynthetic function, it is assumed 
that these genes will remain operational. There may be cases where internal stops, bad starts, or missing stops are actually the result of a real mutation, especially 
in parasitic lineages (Krause, 2008). In most cases, however, ReFernment annotations should reflect real RNA editing sites. A major limitation of ReFernment is that the 
annotations it produces are only as good as the annotations it is provided. If a gene annotation is frameshifted, if there are assembly errors, or if an annotation has 
the incorrect start and stop sites ReFernment will not know. In other words, ReFernment is not a substitute for manually checking gene annotations, nor is ReFernment a 
fix for sloppy annotation. If there are more than 5 detected internal stops in a gene ReFernment will suggest to the user that they manually check that gene. There are 
cases where real genes have more than 5 RNA edited internal stops, but these are relatively rare, so users should use best judgement.</p>

<h2>How to use it</h2>

<p>ReFernment requires both a GenBank flat file and a GFF3 file (without sequence), to produce annotations for each genome. ReFernment takes as input the following:</p>

<ul>
<li><code>gbFolderPath</code>: directory containing the GB files</li>
<li><code>gffFolderPath</code>: directory containing the GFF files</li>
<li><code>outputFolderPath</code>: the directory you would like to output the newly generated GenBank flat files</li>
<li><code>genomes</code>: the names of the files being used</li>
</ul>

<p>The <code>gbFolderPath</code> and <code>gffFolderPath</code> can refer to the the same directory, but unless you are <em>totally comfortable</em> with ReFernment overwriting the original GB files, 
then <code>outputFolderPath</code> should refer to a directory separate from the other two. When preparing to use ReFernment you should keep the names of the GB files and GFF files 
identical, except for their extensions. For example <code>Hemionitis_subcordata.gb</code> and <code>Hemionitis_subcordata.gff</code>. This makes it easy for ReFernment to loop through many 
files all at once if you have large numbers of plastomes that need to be annotated. Along those lines, <code>genomes</code> should not contain the file extension. ReFernment will 
figure add the file extensions later.</p>

<p>Below we want ReFernment to annotate the plastomes of Asplenium pekinense and Woodwardia unigemmata. To start, we open an R console (either in your preferred command line interface or in RStudio), 
load ReFernment into the workspace. Then we declare a vector named <code>genomes</code> which contains all of the plastomes which you would like to annotate. </p>

<pre><code class="r">install.packages(&quot;path/to/ReFernment/&quot;)
#add the above command if this is your first time using ReFernment
library(&quot;ReFernment&quot;)
genomes &lt;- c(&quot;Asplenium_pek&quot;, &quot;Woodwardia_uni&quot;)
</code></pre>

<p>Next, we&#39;ll declare vectors containing the paths to both the input folders (<code>gb</code> and <code>gff</code>)</p>

<pre><code class="r">gbFolder &lt;- &quot;C:/Users/Me/Location/Of/GB/Files/&quot;

gffFolder &lt;- &quot;C:/Users/Me/Location/Of/gff/Files/&quot;

outputFolder &lt;- &quot;C:/Users/Me/Location/Of/Output/Folder/&quot;
</code></pre>

<p>Note that if we wanted, <code>gbFolder</code> and <code>gffFolder</code> could refer to the same directory</p>

<pre><code class="r">gbFolder &lt;- &quot;C:/Users/Me/Location/Of/input/Files/&quot;

gffFolder &lt;- &quot;C:/Users/Me/Location/Of/input/Files/&quot;

outputFolder &lt;- &quot;C:/Users/Me/Location/Of/Output/Folder/&quot;
</code></pre>

<p>Finally, we simply call the <code>ReFernment</code> function and wait for it to finish. This can take several minutes if you have a large number of plastomes. </p>

<pre><code class="r">ReFernment(gbFolder, gffFolder, outputFolder, genomes)
</code></pre>

<p>As ReFernment runs, it may produce the following warning message:</p>

<pre><code>There are a high number of edited Stops ( [number] ) in [geneName] manually check to make sure frame is correct
</code></pre>

<p>This message is intended to warn the user of two common problems:</p>

<ol>
<li>potential assembly errors that would result in an entire gene appearing to be frameshifted</li>
<li>annotation errors where the proper frame for a gene was not selected to begin with</li>
</ol>

<p>The warning will appear if a given coding sequence has more than 5 internal stops. There are, of course, cases where this happens and it is not the result of assembly error, 
but these are relatively rare. Especially if there are more than 10 internal stops the user should take a close look at this coding sequecne to check for assembly error or 
annotation errors. </p>

<h2>Conclusions</h2>

<p>ReFernment offers easy and rapid annotation of RNA edited sites and automatic conceptual translation of amino acid sequences, streamlining the process of GenBank 
submission and saving the user valuable time. </p>

</body>

</html>
